v-next/hardhat/src/internal/cli/error-handler.ts

```diff
@@ -0,0 +1,151 @@
+import util from "node:util";
+
+import {
+  HardhatError,
+  HardhatPluginError,
+} from "@ignored/hardhat-vnext-errors";
+import { isCi } from "@ignored/hardhat-vnext-utils/ci";
+import chalk from "chalk";
+
+import { HARDHAT_NAME, HARDHAT_WEBSITE_URL } from "../constants.js";
+
+/**
+ * The different categories of errors that can be handled by hardhat cli.
+ * Each category has a different way of being formatted and displayed.
+ * To add new categories, add a new entry to this enum and update the
+ *  {@link getErrorWithCategory} and {@link getErrorMessages} functions
+ * accordingly.
+ */
+enum ErrorCategory {
+  HARDHAT = "HARDHAT",
+  PLUGIN = "PLUGIN",
+  COMMUNITY_PLUGIN = "COMMUNITY_PLUGIN",
+  OTHER = "OTHER",
+}
+
+type ErrorWithCategory =
+  | {
+      category: ErrorCategory.HARDHAT;
+      categorizedError: HardhatError;
+    }
+  | {
+      category: ErrorCategory.PLUGIN;
+      categorizedError: HardhatError;
+    }
+  | {
+      category: ErrorCategory.COMMUNITY_PLUGIN;
+      categorizedError: HardhatPluginError;
+    }
+  | {
+      category: ErrorCategory.OTHER;
+      categorizedError: unknown;
+    };
+
+/**
+ * The different messages that can be displayed for each category of errors.
+ *  - `formattedErrorMessage`: the main error message that is always displayed.
+ *  - `showMoreInfoMessage`: an optional message that can be displayed to
+ * provide more information about the error. It is only displayed when stack
+ * traces are hidden.
+ *  - `postErrorStackTraceMessage` an optional message that can be displayed
+ * after the stack trace. It is only displayed when stack traces are shown.
+ */
+interface ErrorMessages {
+  formattedErrorMessage: string;
+  showMoreInfoMessage?: string;
+  postErrorStackTraceMessage?: string;
+}
+
+/**
+ * Formats and logs error messages based on the category the error belongs to.
+ *
+ * @param error the error to handle. Supported categories are defined in
+ * {@link ErrorCategory}.
+ * @param print the function used to print the error message, defaults to
+ * `console.error`. Useful for testing to capture error messages.
+ */
+export function printErrorMessages(
+  error: unknown,
+  print: (message: string) => void = console.error,
+): void {
+  const showStackTraces =
+    process.argv.includes("--show-stack-traces") ||
```

Comment:
I'd pass this from `main`, so that we keep everything argv parsing there. It's a bit unexpected here.

Reply:
Done in https://github.com/NomicFoundation/hardhat/pull/5455/commits/8111de5583807c3c6e62fe51030f313c746cf329